### PR TITLE
(3 of 2) of the rest of the new bytecode emitter (yes, this PR is Part 3 of 2)

### PR DESCRIPTION
**Pull request by magarciaEPFL**
Supersedes https://github.com/scala/scala/pull/2889 and incorporates feedback from there.

This PR completes the new bytecode emitter, including building and writing class files in parallel, by building upon #2863

Before writing class files, a modicum of code tightening transformations are necessary: unreachable code elimination, which in turn calls for cleanup of labels and exception entries made redundant by unreachable code elimination.

In this PR, each such transformation is introduced in a commit with tests for the transformation in isolation. Another commit also in this PR tests each such transformation in connection with the bytecode emitter.

### CODE REVIEW

**Comment by gkossakowski** [comment]
@magarciaEPFL: The commit message for 4cbd067 is wrong. You took it from my commit which made sense at the time but after squashing both commits it's misleading because it suggests that it cleans up _existing_ JUnit tests. In fact, it rewrites unit tests written in partest as JUnit tests.

**Comment by magarciaEPFL** [comment]
@gkossakowski The commit message for https://github.com/scala/scala/commit/4cbd067 , to recap: "Cleanup JUnit tests for GenBCode" reads quite understandable to me, my reading is "Cleaned up JUnit tests ..." (because the result, whether those tests were originally partest or not, are JUnit tests, the commit message does not purport to describe what those tests were before). Now that you mention it, another reading could be "Cleanup of JUnit tests ...".

I guess that by adding words (as in both cases) even more readings are possible. I don't see how any of those readings can be more "unique" than the others. My reading is as good as yours.

**Comment by magarciaEPFL** [comment]
Can we please turn the page on this PR? It's been ultra-reviewed, it's been ready to be merged for days and weeks now, and so on. Can we just turn the page, merge it, and move on?

File: src/asm/scala/tools/asm/MethodWriter.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
  * @author Eric Bruneton
  * @author Eugene Kuleshov
  */
-class MethodWriter extends MethodVisitor {
```
Comment (adriaanm) [inline src/asm/scala/tools/asm/MethodWriter.java:40]: So, what about our agreement regarding not forking ASM?

Comment (adriaanm) [inline src/asm/scala/tools/asm/MethodWriter.java:40]: Have you talked to the ASM authors regarding merging this change upstream?

**Comment by adriaanm** [comment]
@magarciaEPFL I don't see having to "ultra-review" PRs as a good thing. I'd rather get PRs that follow the rules that we all agreed upon, and all code by, so that reviewing is a smooth and efficient experience for both reviewer and submitter. I also don't see how advice regarding proper testing is showing off. Please check your attitude at the door.

**Comment by adriaanm** [comment]
@magarciaEPFL, I truly appreciate all the time you've spent on this PR, but let's not forget the considerable effort and patience that's gone into reviewing it, and the maintenance futures to go with merging this. We'll spend as much time as it takes to polish this to the point that it can go in without having to live in fear of the day the technical debt collectors will come knocking.

tl;dr: Just a little extra push and I'm confident this will land in M6 next month.

**Comment by magarciaEPFL** [comment]
The pros and cons of waiting for the ASM authors to approve the changes proposed here were discussed in PR https://github.com/scala/scala/pull/2711 

Pros
- saving a few minutes whenever a new release of ASM comes along (the time it takes to apply the patches proposed here)

Cons
- the ASM authors become the owners and maintainers of functionality the new optimizer relies on (eg, they can remove it as they please)

The proposed customization of ASM comprises:
- getters in MethodWriter (part of this PR, https://github.com/magarciaEPFL/scala/commit/f482aa52342529fa3419de9c0fb81d489092c047 )
- more precise abstract-values in ASM dataflow-analysis (upcoming PR, https://github.com/magarciaEPFL/scala/commit/88e0f8e4cacf0471eee1c4a337847cacea304876 )
- helper methods in ASM dataflow-analysis (upcoming PR, https://github.com/magarciaEPFL/scala/commit/201fd7aff5fb9337dd8c39c066cf1206d9d79383 )

To the untrained eye, those customizations amount to forking. To those familiar with ASM, they are well-engineered and focused improvements that pay off big time. In fact, I was expecting the ASM connoisseurs of your team to jump in and say: "Hey, that extension to the iterative dataflow framework in ASM is easy cake! I should have thought about it before!".

For example, all intra-method optimizations (upcoming PR) are about iterative DFAs. Maintaining that code requires the same level of ASM know-how as needed to maintain the ASM customizations proposed here. That's why I'm surprised nobody vetoes intra-method optimizations yet "ASM forking" is being vetoed. Something doesn't add up in the veto decision. Either veto none or all of them.

I wanted to make crystal clear the alternatives, for the record.

**Comment by retronym** [comment]
The build kitteh was misbehaving and jobs from this PR were being built over and over. I've tried to manually cancel them in Jenkins to clear the congestion. Please disregard the latest build failures.

**Comment by adriaanm** [comment]
> To the untrained eye, those customizations amount to forking. To those familiar with ASM, they are well-engineered and focused improvements that pay off big time. In fact, I was expecting the ASM connoisseurs of your team to jump in and say: "Hey, that extension to the iterative dataflow framework in ASM is easy cake! I should have thought about it before!".

Whether something is a fork does not depend on the quality of the code. Whether the patch is accepted upstream does depend on that, and you have thus made a great case against the only contra in your list of submitting this to the ASM project. (And again, while I agree one small change is not a big burden, it's the slippery slope it puts us on. If we deem it necessary to fork ASM, we'll have to do so outside of the scala repo, and set up all the infrastructure around it.)

**Comment by adriaanm** [comment]
I have no idea why the IDE integration fails, but not seeing those errors in the other PRs. Perhaps this needs a rebase? Closing for now.

**Comment by dragos** [comment]
@adriaanm, yes, it needs a rebase (it looks like dbuild can't find the metadata to infer module versions)
